src/Language/DifferentialDatalog/Compile.hs

```diff
@@ -1423,6 +1428,66 @@ mkRelId2NameC =
     $$ "    RELIDMAPC.get(&rid).copied()"
     $$ "}"
 
+-- The original relation name is retrieved from the 'original="name"' attribute
+-- if present.  Validation ensures that at most one such attribute exists and
+-- that is has a string argument.
+-- If no attribute exists the relation name is returned
+origRelName :: Relation -> Doc
+origRelName rel = let attrs = map attrVal $ filter ((== "original") . name) $ relAttrs rel in
+                  case attrs of
+                      [E (EString _ str)] -> pp $ str
+                      -- The only remaining case should be an empty list
+                      _                   -> pp $ relName rel
+
+-- Generates a function that converts relation names into the original relation names
+-- (obtained via the 'original' attribute)
+mkRelName2OrigName :: DatalogProgram -> Doc
+mkRelName2OrigName d =
+    "pub fn rel_name2orig_name(rname: &str) -> ::std::option::Option<&'static str> {" $$
+    "   match rname {"                                        $$
+    (nest' $ nest' $ vcat $ entries)                          $$
+    "       _  => None"                                       $$
+    "   }"                                                    $$
+    "}"
+    where
+    entries = map mkrel $ M.elems $ progRelations d
+    mkrel :: Relation -> Doc
+    mkrel rel = (doubleQuotes $ pp $ name rel) <+> "=> ::core::option::Option::Some(\"" <> pp (origRelName rel) <> "\"),"
+
+-- Generates a function that converts relation names into the original relation names
+-- (obtained via the 'original' attribute)
+mkRelName2OrigNameC :: DatalogProgram -> Doc
+mkRelName2OrigNameC d =
+    "#[cfg(feature = \"c_api\")]"                                                                   $$
+    "pub fn rel_name2orig_cname(rname: &str) -> ::std::option::Option<&'static ::std::ffi::CStr> {" $$
+    "   match rname {"                                        $$
+    (nest' $ nest' $ vcat $ entries)                          $$
+    "       _  => None"                                       $$
+    "   }"                                                    $$
+    "}"
+    where
+    entries = map mkrel $ M.elems $ progRelations d
+    mkrel :: Relation -> Doc
+    mkrel rel = (doubleQuotes $ pp $ origRelName rel) <+> "=>" <+>
+            "Some(::std::ffi::CStr::from_bytes_with_nul(b\"" <> pp (name rel) <> "\\0\")" <+>
```

Comment:
I think it'd be 100% valid to write `#[original = "foobar\0"]` so I think we should just add a quick error check that gives a compile error if the user tries it, that's much better than a panic deep within generated code deep within a user's application imo

Reply:
This pattern is used a lot in the rust code, so perhaps we should actually file an issue for it rather than fix it here.

Reply:
Filed this: https://github.com/vmware/differential-datalog/issues/997

Reply:
I mean, both can be done. This code is being touched right now, may as well fix it